File: app/src/main/java/com/wolcano/musicplayer/music/content/managers/SoundManager.java
package com.wolcano.musicplayer.music.content.managers;

import android.content.Context;
import android.media.AudioManager;

import com.wolcano.musicplayer.music.provider.RemotePlay;

import static android.content.Context.AUDIO_SERVICE;

public class SoundManager implements AudioManager.OnAudioFocusChangeListener {

    private AudioManager audioManager;
    private boolean isPause;
    private Context context;

    public SoundManager(Context context) {
        this.context = context;
        audioManager = (AudioManager) context.getSystemService(AUDIO_SERVICE);
    }

    @Override
    public void onAudioFocusChange(int focusChange) {
        switch (focusChange) {
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK:
                RemotePlay.get().getMediaPlayer().setVolume(0.5f, 0.5f);
                break;

            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                RemotePlay.get().pauseRemotePlay(context.getApplicationContext(),false);
                isPause = true;
                break;
            case AudioManager.AUDIOFOCUS_LOSS:
                RemotePlay.get().pauseRemotePlay(context.getApplicationContext());
                break;
            case AudioManager.AUDIOFOCUS_GAIN:
                if (isPause) {
                    RemotePlay.get().startRemotePlay(context.getApplicationContext());
                }
                RemotePlay.get().getMediaPlayer().setVolume(1f, 1f);
                isPause = false;
                break;

        }
    }



    public boolean reqSoundFocus() {
        return audioManager.requestAudioFocus(this, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN)
                == AudioManager.AUDIOFOCUS_REQUEST_GRANTED;
    }

    public void abSoundFocus() {
        audioManager.abandonAudioFocus(this);
    }

}

File: app/src/main/java/com/wolcano/musicplayer/music/content/managers/SoundManager.kt
package com.wolcano.musicplayer.music.content.managers

import android.content.Context
import android.media.AudioManager
import android.media.AudioManager.OnAudioFocusChangeListener
import com.wolcano.musicplayer.music.provider.RemotePlay.getMediaPlayer
import com.wolcano.musicplayer.music.provider.RemotePlay.pauseRemotePlay
import com.wolcano.musicplayer.music.provider.RemotePlay.startRemotePlay

class SoundManager(private val context: Context) : OnAudioFocusChangeListener {

    private var audioManager: AudioManager? = null
    private var isPause = false

    init {
        audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    }

    override fun onAudioFocusChange(focusChange: Int) {
        when (focusChange) {
            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> getMediaPlayer()!!.setVolume(0.5f, 0.5f)
            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
                pauseRemotePlay(context.applicationContext, false)
                isPause = true
            }
            AudioManager.AUDIOFOCUS_LOSS -> pauseRemotePlay(context.applicationContext)
            AudioManager.AUDIOFOCUS_GAIN -> {
                if (isPause) {
                    startRemotePlay(context.applicationContext)
                }
                getMediaPlayer()?.setVolume(1f, 1f)
                isPause = false
            }
        }
    }

    fun requestAudioFocus(): Boolean {
        return (audioManager?.requestAudioFocus(
            this,
            AudioManager.STREAM_MUSIC,
            AudioManager.AUDIOFOCUS_GAIN
        )
                == AudioManager.AUDIOFOCUS_REQUEST_GRANTED)
    }

    fun abandonAudioFocus() {
        audioManager?.abandonAudioFocus(this)
    }

}